include/envoy/upstream/cluster_manager.h

```diff
@@ -309,6 +328,32 @@ class ClusterManager {
   virtual const ClusterRequestResponseSizeStatNames&
   clusterRequestResponseSizeStatNames() const PURE;
   virtual const ClusterTimeoutBudgetStatNames& clusterTimeoutBudgetStatNames() const PURE;
+
+  /**
+   * Request an on-demand discovery of a cluster with a passed name. Passed ODCDS may be used to
+   * perform the discovery process in the main thread if there is no discovery going on for this
+   * cluster. The passed callback will be invoked when the cluster is added and warmed up. It is
+   * expected that the callback will be destroyed when it is invoked. To cancel the discovery,
+   * destroy the returned handle and the callback.
+   *
+   * This function is thread-safe.
+   *
+   * @param weak_odcds is a weak pointer to the ODCDS that will be used for discovery.
+   * @param name is the name of the cluster to be discovered.
+   * @param callback will be called when the discovery is finished.
+   * @return the discovery process handle
+   */
+  virtual ClusterDiscoveryCallbackHandlePtr
+  requestOnDemandClusterDiscovery(OdCdsApiWeakPtr weak_odcds, const std::string& name,
```

Comment:
It's odd to pass a weak_ptr here. I would pass a shared_ptr since the ocds should exist at this point and allow the callee to convert to weak_ptr as needed.

Reply:
Yeah, will pass a shared_ptr for odcds here. Now looking at the code, there are probably some corner cases with the weak_ptr, so change to shared_ptr should make it simpler and less cornery. But I'll keep the weak_ptr for callback.